### PR TITLE
ocvsmd as cyphal node (heartbeat & get info)

### PR DESCRIPTION
**Pull request by serges147**
Screenshot from 2024-12-18 19-19-30

- Added `EpollSingleThreadedExecutor` (later there will be kqueue version as well)
- Added platform stuff (error, udp media & sockets)
- Added main `Application` class:
  - implemented super loop
  - created libcyphal UDP transport (for now with hard coded `127.0.0.1` address and node_id=`7` - will be addressed in the next PR about config)
  - created libcyphal presentation and node entities
  - brought up "GetInfo" server with ...
    - name: `org.opencyphal.ocvsmd`
    - sw ver: `0.1`
    - vcs_rev: < short git commit >
    - unique: < randomly generated > (for now not persisted; later will be in the config)
- Added submodules:
  - `libcyphal`, `CETL`, `libudpard`
  - `nunavut`, `pydsdl`, `public_regulated_data_types`
- Updated README.md with new build instructions

Also:
- LLVM 15 -> 18
- "Ninja Multi-Config" as cmake generator
- CMakePresets

### CODE REVIEW

**Review by pavel-kirienko — APPROVED** [review]

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	path = submodules/libudpard
+	url = https://github.com/OpenCyphal/libudpard
+	branch = v2
+[submodule "submodules/nunavut"]
+	path = submodules/nunavut
+	url = https://github.com/OpenCyphal/nunavut.git
+	branch = 3.0.preview
+[submodule "submodules/pydsdl"]
+	path = submodules/pydsdl
+	url = https://github.com/OpenCyphal/pydsdl.git
```
[the remark this answers is not in the document]
Comment (serges147) [reply .gitmodules:14]: I did it exactly the same as Scott just recently did in his "kqueue" [darwin](https://github.com/OpenCyphal-Garage/demos/tree/darwin) branch. I assumed that way is the latest and greatest one. @thirtytwobits Scott?

File: src/daemon/engine/application.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+{
+namespace daemon
+{
+namespace engine
+{
+
+class Application
+{
+public:
+    CETL_NODISCARD cetl::optional<std::string> init();
```
[the remark this answers is not in the document]
Comment (serges147) [reply src/daemon/engine/application.hpp:32]: We probably can, I decided not to spend time on this. Honestly, I don't see big difference for this specific case. Also, I'm under time pressure, so skipping some maybe better practisise in favor of speed. There are no even unit tests currently. My focus on functionality.

File: src/daemon/engine/platform/posix_platform_error.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    virtual ~PosixPlatformError()                                = default;
+    PosixPlatformError(const PosixPlatformError&)                = default;
+    PosixPlatformError(PosixPlatformError&&) noexcept            = default;
+    PosixPlatformError& operator=(const PosixPlatformError&)     = default;
+    PosixPlatformError& operator=(PosixPlatformError&&) noexcept = default;
+
+    // MARK: IPlatformError
+
+    std::uint32_t code() const noexcept override
```
[the remark this answers is not in the document]
Comment (serges147) [reply src/daemon/engine/platform/posix_platform_error.hpp:39]: Yes, errno. Not sure about name. What exactly is your proposal.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -18,23 +17,19 @@ jobs:
...
-          -B ${{ github.workspace }}/build
-          -DCMAKE_BUILD_TYPE=Debug
-          -DCMAKE_CXX_COMPILER=${{ matrix.cxx-compiler }}
-      - working-directory: ${{github.workspace}}/build
-        run: |
-          make VERBOSE=1
-          make test
+      - run: |
+          cmake --preset OCVSMD-Debian -DCMAKE_C_COMPILER=${{ matrix.c-compiler }} -DCMAKE_CXX_COMPILER=${{ matrix.cxx-compiler }}
+          cmake --build --preset OCVSMD-Debian-Debug
```
[the remark this answers is not in the document]
Comment (serges147) [reply .github/workflows/tests.yml:32]: Debian - for Ubuntu; Darwin - will be for MacOS. On Ubuntu we gonna use "epoll"-based executor. On macOS - "kqueue"-based one (near future pr).

File: src/daemon/engine/platform/udp/udp_media.hpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    cetl::pmr::memory_resource& getTxMemoryResource() override
+    {
+        return tx_mr_;
+    }
+
+    // MARK: Data members:
+
+    cetl::pmr::memory_resource& general_mr_;
+    libcyphal::IExecutor&       executor_;
+    std::string                 iface_address_;
```
[the remark this answers is not in the document]
Comment (serges147) [reply src/daemon/engine/platform/udp/udp_media.hpp:87]: The same reason as I described for `Application::init`, namely to save time, and quicker achieve functionality.

File: src/daemon/engine/platform/udp/udp.c
[the remark this answers is not in the document]
Comment (serges147) [reply src/daemon/engine/platform/udp/udp.c:1]: probably, but I do quicker solution - just copy/paste for now

File: src/daemon/engine/platform/posix_platform_error.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply src/daemon/engine/platform/posix_platform_error.hpp:39]:
```diff
-     std::uint32_t code() const noexcept override
+     std::uint32_t errno() const noexcept override
```

File: .github/workflows/tests.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply .github/workflows/tests.yml:32]: What I'm saying is that the first invocation of cmake uses `OCVSMD-Debian` while the second is `OCVSMD-Debian-Debug`, is this intentional?

[the remark this answers is not in the document]
Comment (serges147) [reply .github/workflows/tests.yml:32]: Yes, intentional. CMake works in two stages:
- first is configuration - b/c we now use "ninja mutli config" generator there is no notion of Debug/Release.
- second is build - this is where you decide which config (Debug or Release) to build

File: src/daemon/engine/platform/posix_platform_error.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (serges147) [reply src/daemon/engine/platform/posix_platform_error.hpp:39]: I'll rename it in the next pr

[the remark this answers is not in the document]
Comment (serges147) [reply src/daemon/engine/platform/posix_platform_error.hpp:39]: but note that this `code()` is actually override of `IPlatformError::code`. Original idea of such naming (as `code`) was b/c implementation could use something not necessarily `errno`-based as "code" of error.

File: src/daemon/engine/application.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply src/daemon/engine/application.hpp:32]: Okay; please add some TODO comments then around this init method and also the use of heapy string mentioned earlier

File: src/daemon/engine/platform/posix_platform_error.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply src/daemon/engine/platform/posix_platform_error.hpp:39]: Ah yes, it's a generic interface. Then we need at least a comment explaining that this is `errno`, in this specific implementation.

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (thirtytwobits) [reply .gitmodules:14]: I'm trying to simplify integrating Nunavut by using the submodule baked into this repo.
See this example -> https://github.com/OpenCyphal/nunavut/blob/3.0.preview/docs/cmake/CMakeLists.txt
and this new build step -> https://github.com/OpenCyphal/nunavut/blob/3.0.preview/pydsdl_version_check.py
which makes sure that, in future releases of Nunavut, the python dependency is in sync with the git submodule.

In the CMakeLists example you'll see the integration is greatly simplified in that the user does not need to create a suitable python environment nor do they have to maintain the correct pydsdl dependency. With one gitpull they get the dependencies they need and one `find_package` call (plus the find python call Nunavut does automatically) and they've integrated Nunavut into their build. The only way we could make this simpler is if we had a native version of Nunavut that didn't rely on Python at all.

File: src/daemon/engine/platform/udp/udp_media.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (thirtytwobits) [reply src/daemon/engine/platform/udp/udp_media.hpp:87]: I'm okay with this. Let's be frugal with memory but we don't have to be as pedantic as with libcyphal itself given that this is targeted for a Linux computer.

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (thirtytwobits) [reply .gitmodules:14]: To be clear, this isn't yet right but we can fix it up later. Let's not block on it.

**Review by thirtytwobits — APPROVED** [review]

**Pull request merged**